x/programs/rust/wasmlanche-sdk/src/state.rs

```diff
@@ -159,87 +186,65 @@ impl Key {
     }
 }
 
-macro_rules! ffi_linker {
-    ($mod:literal, $link:literal, $caller:ident, $key:ident) => {
-        #[link(wasm_import_module = $mod)]
-        extern "C" {
-            #[link_name = $link]
-            fn ffi(caller: CPointer, key: CPointer) -> i32;
-        }
-
-        let $caller = to_ffi_ptr($caller.id())?;
-        let $key = to_ffi_ptr($key)?;
-    };
-    ($mod:literal, $link:literal, $caller:ident, $key:ident, $value:ident) => {
-        #[link(wasm_import_module = $mod)]
-        extern "C" {
-            #[link_name = $link]
-            fn ffi(caller: CPointer, key: CPointer, value: CPointer) -> i32;
-        }
-
-        let $caller = to_ffi_ptr($caller.id())?;
-        let $key = to_ffi_ptr($key)?;
-        let $value = to_ffi_ptr($value)?;
-    };
+#[derive(BorshSerialize)]
+pub(super) struct PutArgs {
+    pub(super) caller: Program,
+    pub(super) key: Vec<u8>,
+    pub(super) bytes: Vec<u8>,
 }
 
-macro_rules! call_host_fn {
-    (
-        wasm_import_module = $mod:literal
-        link_name = $link:literal
-        args = ($caller:ident, $key:ident)
-    ) => {{
-        ffi_linker!($mod, $link, $caller, $key);
-
-        unsafe { ffi($caller, $key) }
-    }};
-
-    (
-        wasm_import_module = $mod:literal
-        link_name = $link:literal
-        args = ($caller:ident, $key:ident, $value:ident)
-    ) => {{
-        ffi_linker!($mod, $link, $caller, $key, $value);
-
-        unsafe { ffi($caller, $key, $value) }
-    }};
```

Comment:
Don't worry @iFrostizz, this will come back, it'll just look a little different

Reply:
Can you not just replace all of this with

```rust
macro_rules! call_host_fn {
    (
        wasm_import_module = $mod:literal
        link_name = $link:literal
        args = $value:ident
    ) => {{
        #[link(wasm_import_module = $mod)]
        extern "C" {
            #[link_name = $link]
            fn ffi(ptr: *const u8, len: usize) -> usize;
        }

        unsafe { ffi($value.as_ptr(), $value.len()) }
    }};
}
```

Then call it with

```rust
    /// Persists the bytes at key on the host storage.
    pub(super) fn put_bytes(bytes: &[u8]) -> Result<(), Error> {
        match call_host_fn! {
            wasm_import_module = "state"
            link_name = "put"
            args = bytes
        } {
            0 => Ok(()),
            _ => Err(Error::Write),
        }
    }
```

Reply:
Yes! I want to figure out a better way to handle the return type instead of just hardcoding a size and then casting, so I decided to leave this for another PR. I want to figure out a way to normalize the return type for all host-calls. it's easy for most calls, but distinguishing between no-value and an error at an existing key becomes a little harder